### ISSUE TITLE
Problems with build.jl from BinaryBuilder

### ISSUE THREAD

**Issue by lobingera** [JuliaPackaging/BinaryProvider.jl#176]
I'm trying to integrate from here
https://github.com/lobingera/Cairo.jl/blob/bb2/deps/build.jl
and get longer list of
```
julia> BinaryProvider.locate(products[1],verbose=true,isolate=true)
[ Info: Found a valid dl path libcairo-script-interpreter.so while looking for libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo-script-interpreter.so matches our search criteria of libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo-script-interpreter.so cannot be dlopen'ed
[ Info: Found a valid dl path libcairo-script-interpreter.so.2 while looking for libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo-script-interpreter.so.2 matches our search criteria of libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo-script-interpreter.so.2 cannot be dlopen'ed
[ Info: Found a valid dl path libcairo-script-interpreter.so.2.11400.12 while looking for libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo-script-interpreter.so.2.11400.12 matches our search criteria of libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo-script-interpreter.so.2.11400.12 cannot be dlopen'ed
[ Info: Found a valid dl path libcairo.so while looking for libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo.so matches our search criteria of libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo.so cannot be dlopen'ed
[ Info: Found a valid dl path libcairo.so.2 while looking for libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo.so.2 matches our search criteria of libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo.so.2 cannot be dlopen'ed
[ Info: Found a valid dl path libcairo.so.2.11400.12 while looking for libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo.so.2.11400.12 matches our search criteria of libcairo
[ Info: /home/lobi/.julia/dev/Cairo/deps/usr/lib/libcairo.so.2.11400.12 cannot be dlopen'ed
[ Info: Could not locate libcairo inside /home/lobi/.julia/dev/Cairo/deps/usr/lib
```

although,
```
lobi@orange4:~/.julia/dev/Cairo/deps/usr/lib$ ls
bin          libcairo-script-interpreter.la             logs
cairo        libcairo-script-interpreter.so             man
cmake        libcairo-script-interpreter.so.2           manifests
downloads    libcairo-script-interpreter.so.2.11400.12  pkgconfig
include      libcairo.so                                python2.7
lib          libcairo.so.2                              share
libcairo.la  libcairo.so.2.11400.12
```

build.jl was borrowed from
 https://github.com/giordano/Yggdrasil/releases/download/Cairo-v1.14.12/build_Cairo.v1.14.12.jl

**Comment by staticfloat** [comment]
@giordano investigated this and discovered that it was due to an implicit ordering between libraries; if libA depends on libB, then including the `build.jl` for libB before that of libA would cause libB to not be dlopen'able, which would manifest in problems like above.  By permuting the ordering of `build.jl` files being included, he fixed the issues.

This will be fixed for Julia 1.3+ as we are able to take advantage of the Artifacts and JLL package systems, which will impose an implicit ordering upon library open order.

**Comment by lobingera** [comment]
@staticfloat I see - can you please document this?
